File: src/macrotone.c
/*
  macrotone
  Copyright (C) 2021 Vlad-Stefan Harbuz <vlad@vladh.net>
  GNU General Public License v3 (https://www.gnu.org/licenses). Absolutely no warranty.
*/

#include <SDL2/SDL.h>
#include <SDL2/SDL_audio.h>

#include "types.h"
#include "audio.h"
#include "theory.h"


void run_main_loop(bool32 *should_stop, AudioState *audio_state) {
  uint32 const pitches_capacity = 128;
  Pitch pitches[pitches_capacity];
  uint32 n_pitches = make_pythagorean_scale(pitches, pitches_capacity, THEORY_A_HZ / 2.0f);
  for (uint32 idx = 0; idx < n_pitches; idx++) {
    audio_play_note_blocking(audio_state, pitches[idx], 500.0f);
  }
  SDL_Delay(3000);
}


int main(int argc, char *argv[]) {
  bool32 should_stop = false;
  AudioState audio_state = {0};
  SDL_AudioSpec target_spec = {
    .freq = AUDIO_SAMPLE_RATE,
    .format = AUDIO_S16SYS,
    .channels = 1,
    .samples = 2048,
    .callback = audio_fill_buffer,
    .userdata = &audio_state,
  };
  if (!audio_init(target_spec)) {
    goto cleanup_end;
  }

  // Start playing audio.
  SDL_PauseAudio(0);

  while (!should_stop) {
    run_main_loop(&should_stop, &audio_state);
  }

  audio_destroy();
cleanup_end:
  return EXIT_SUCCESS;
}

File: macrotone.go
/*
  macrotone
  Copyright (C) 2021 Vlad-Stefan Harbuz <vlad@vladh.net>
  GNU General Public License v3 (https://www.gnu.org/licenses). Absolutely no warranty.
*/

package main

import (
  "log"

  "github.com/veandco/go-sdl2/sdl"

  "net.vladh/macrotone/internal/audio"
  "net.vladh/macrotone/internal/theory"
)


func runMainLoop(audioState *audio.AudioState) {
  log.Println("Tick")

  pitches := theory.MakePythagoreanMajorScale(440.0)

  for _, pitch := range pitches {
    audio.PlayNoteBlocking(audioState, pitch, 500.0)
  }

  sdl.Delay(1000.0)
}


func main() {
  log.SetFlags(0)
  log.Println("Hello")

  var audioState audio.AudioState

  if err := audio.Init(&audioState); err != nil {
    log.Println(err)
    return
  }
  defer audio.Destroy()

  // Start playing audio
  sdl.PauseAudio(false)

  log.Println("Pythagorean interval ratios:")
  var i float64 = 0
  for i = -6; i <= 6; i++ {
    num, denom := theory.MakePythagoreanIntRatio(i)
    log.Printf("%2.0f: %2.2f / %2.2f\n", i, num, denom)
  }

  for {
    runMainLoop(&audioState)
  }
}
